src/libraries/System.Reflection.DispatchProxy/src/System/Reflection/DispatchProxyGenerator.cs

```diff
@@ -170,17 +91,34 @@ private sealed class GeneratedTypeInfo
 
         private sealed class ProxyAssembly
         {
+            // Proxies are requested for a pair of types: base type and interface type.
+            // The generated proxy will subclass the given base type and implement the interface type.
+            // We maintain a cache keyed by 'base type' containing a dictionary keyed by interface type,
+            // containing the generated proxy type for that pair. There are likely to be few (maybe only 1)
+            // base type in use for many interface types.
+            // Note: this differs from Silverlight's RealProxy implementation which keys strictly off the
+            // interface type. But this does not allow the same interface type to be used with more than a
+            // single base type. The implementation here permits multiple interface types to be used with
+            // multiple base types, and the generated proxy types will be unique.
+            // This cache of generated types grows unbounded, one element per unique T/ProxyT pair.
+            // This approach is used to prevent regenerating identical proxy types for identical T/Proxy pairs,
+            // which would ultimately be a more expensive leak.
+            // Proxy instances are not cached. Their lifetime is entirely owned by the caller of DispatchProxy.Create.
+            private readonly Dictionary<Type, Dictionary<Type, GeneratedTypeInfo>> _baseTypeAndInterfaceToGeneratedProxyType = new Dictionary<Type, Dictionary<Type, GeneratedTypeInfo>>();
+
             private readonly AssemblyBuilder _ab;
             private readonly ModuleBuilder _mb;
             private int _typeId;
 
-            private readonly HashSet<string?> _ignoresAccessAssemblyNames = new HashSet<string?>();
+            private readonly HashSet<string> _ignoresAccessAssemblyNames = new HashSet<string>();
             private ConstructorInfo? _ignoresAccessChecksToAttributeConstructor;
 
-            public ProxyAssembly()
+            public ProxyAssembly(AssemblyLoadContext alc)
             {
-                _ab = AssemblyBuilder.DefineDynamicAssembly(new AssemblyName("ProxyBuilder"), AssemblyBuilderAccess.Run);
-                _mb = _ab.DefineDynamicModule("testmod");
+                string? alcName = alc.Name;
+                string name = string.IsNullOrEmpty(alcName) ? $"DispatchProxyTypes.{alc.GetHashCode()}" : $"DispatchProxyTypes.{alcName}";
```

Comment:
> Storing the Default ALC's data in a separate static (outside of the weak table) - to avoid the cost of the weak table for the most common cases.

The weak table costs nothing compared to the cost of the DispatchProxy. I think it is fine to keep this detail simple and store everything in the same weak table.

Reply:
Thank you for your review/suggestions @vitek-karas @jkotas, those are good points

@teo-tsirpanis as suggested above please use the old assembly name "ProxyBuilder" for the Default ALC  for back compatibility

Reply:
Done. I am wondering though, what is the risk of changing the dynamic assembly's name? How could user code depend on this?

Reply:
We've learned the hard way not to assume anything. I agree it sounds weird, but if it's easy for us to maintain the same behavior, we always tend to do that.

Reply:
I see. As a precautionary measure, I renamed the dynamic module back to `testmod`.